### PR TITLE
Feat/burn nft

### PR DESCRIPTION
**Pull request by guru-web3**
## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Issue Number: N/A

## What is the new behavior?

-
-
-

### CODE REVIEW

File: src/controllers/TorusController.ts [hunk 1]
```diff
@@ -631,6 +632,44 @@ export default class TorusController extends BaseController<TorusControllerConfi
...
+    try {
+      const tx = new Transaction();
+      // get the publickey of the NFT
+      const mintPublickey = new PublicKey(NFTtoBurn);
+
+      // signer pub key
+      const signer = new PublicKey(senderAddress);
+
+      // determine the associated token account of the NFT
+      const associatedAddress = await getAssociatedTokenAddress(mintPublickey, signer);
```
Comment (ieow) [inline src/controllers/TorusController.ts:646]: Do not depend on PDA for burning.
get the exact token account address for burning.
It is possible nft is not PDA

File: src/controllers/TorusController.ts [hunk 2]
```diff
@@ -631,6 +632,44 @@ export default class TorusController extends BaseController<TorusControllerConfi
     }
   }
 
+  // burn NFT
+  async burnToken(NFTtoBurn: string, senderAddress: string) {
+    try {
```
Comment (ieow) [inline src/controllers/TorusController.ts:637]: If possible try create a function for generate the burn and close instruction in solanaHelper file
This way we can have tests by simulate the instruction generated

File: src/utils/instructionDecoder.ts [hunk 3]
```diff
@@ -319,81 +322,115 @@ export const decodeInstruction = (instruction: TransactionInstruction): DecodedD
...
-  if (instructions.length === 1 || intrestedInstructionidx > 0) {
-    if (TOKEN_PROGRAM_ID.equals(instructions[intrestedInstructionidx].programId)) {
-      const decoded = decodeTokenInstruction(instructions[intrestedInstructionidx]);
-      // There are transfer and transferChecked type
-      if (decoded.type.includes("transfer")) {
-        const from = new PublicKey(decoded.data.source || "").toBase58();
-        const to = new PublicKey(decoded.data.destination || "").toBase58();
+    // Expect SPL token transfer transaction have only 1 instruction
+    if (instructions.length === 1 || interestedTransferInstructionidx > 0 || interestedBurnInstructionidx > -1) {
+      if (TOKEN_PROGRAM_ID.equals(instructions[interestedTransferInstructionidx].programId)) {
```
Comment (ieow) [inline src/utils/instructionDecoder.ts:373]: Shouldn't the burn instruction index is used to checked ?
another if check should be created

File: src/controllers/TorusController.ts (hunk 1 above)
Comment (metallicalfa2) [reply src/controllers/TorusController.ts:646]: find all token accounts by owner, then filter for the NFT mint address. Its possible that the associatedAddress isn't holding the mint account. 

optionally, you can check the accountInfo for :646 to verify the mint address. If that fails, you can resort to the above.

**Comment by ieow** [comment]
Screenshot 2022-08-02 at 5 45 37 PM

Change the confirm button to Burn and Close
Change the confirm button color to Red
add text Input, so that user required to type symbol/name before they allow to click burn button

**Review by metallicalfa2 — APPROVED** [review]
lgtm

**Pull request merged**